### PR TITLE
Network: Persisted Requests Queue - clear persisted requests after the response

### PR DESCRIPTION
**Pull request by kidroca**
### Details

Additional functionality was added to `Network.js` and `NetworkRequestQueue.js` to better handle persisted requests
- clear requests from persistent storage only after they are completed
- retry persisted request

Added unit tests covering persisted request handling

### Fixed Issues

$ https://github.com/Expensify/App/issues/5987

### Tests

1. Launch the app
2. Go offline
3. Write a few messages
4. Close the app
5. Be online
6. Open the app
7. Observe the messages sent while being offline are correctly sent to the backend

### QA Steps

Follow the steps from the bug report

1. Make sure you are offline.
2. Send some messages to any user.
3. Refresh the page on the web. Close & open the app on mobile.
4. Now you can still see the messages.
5. Now enable internet
6. Shortly (a second after resuming the connection) after refresh or close the app
7. The messages should be posted after the refresh or after you open the app again
  - note: the order of the messages might be incorrect, this PR is not addressing the order

This comment has additional details on steps and timing: https://github.com/Expensify/App/issues/5987#issuecomment-962959408

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

#### Web

https://user-images.githubusercontent.com/12156624/144221544-5308bdb4-902b-4f5f-acb1-9564ac637777.mp4

#### Mobile Web

https://user-images.githubusercontent.com/12156624/144224565-d756ee4b-d2d4-4c50-b35e-fb5b1ebce8c9.mp4

#### Desktop

https://user-images.githubusercontent.com/12156624/144230141-1c63ea16-429f-4544-b949-4cbd1a7c968f.mov

#### iOS

https://user-images.githubusercontent.com/12156624/144236936-5b52e27b-cb66-4cde-8272-17073c9fa35d.mov

#### Android

https://user-images.githubusercontent.com/12156624/144220867-0adf287f-c63d-4220-8a94-0a2fa5ea89c5.mp4

### CODE REVIEW

File: src/libs/Network.js
Comment (kidroca) [inline src/libs/Network.js:67]: We can have remaining requests when 
- the request failed for some reason and is not removed from the queue - as seen in the catch block
- we became offline again, which would again lead to the catch block above. In that case the queue would still stop and will be retriggered once we're back online, more requests can be added to it in the meantime

Not sure if we should capture that in a comment or something

File: src/libs/Network.js
Comment (kidroca) [inline src/libs/Network.js:331]: This is now exported from `CONST.NETWORK.PROCESS_REQUEST_DELAY_MS` and usages updated

File: src/libs/actions/NetworkRequestQueue.js
Comment (kidroca) [inline src/libs/actions/NetworkRequestQueue.js:36]: Retry counts are not persisted to disk - no need to 
If a request is retried 5 times then the app quit, when the app is launched again the request will have a fresh count of up to 10 attempts, and not only the remaining 5

Might be an edge case but I think it would be best to have a retry limit in case some corrupt data was persisted and it cannot be sent no matter how many times we try

**Review by kidroca — COMMENTED** [review]
Added some highlights on the changes

**Comment by johnmlee101** [comment]
@roryabraham do you mind co-reviewing this with me?

File: src/libs/Network.js
Comment (kidroca) [reply src/libs/Network.js:14]: The name comes from this usagage and readability reasons 
This method is **starting** something rather than making it **not paused**

```js
function flushPersistedRequestsQueue() {
    if (persistedRequestsQueueRunning) {
        return;
    }

    persistedRequestsQueueRunning = true;
```

vs

```js
function flushPersistedRequestsQueue() {
    if (!persistedRequestsQueuePaused) {
        return;
    }

   persistedRequestsQueuePaused = false;
```

**running** is the "special case" for this queue, while **paused** is the special case for the regular queue 

I'm reluctant to refactor `isQueuePaused` to `isMainQueueRunning` it would change usages and might affect their intent as well, also it kind a sound like we're online too

**Comment by kidroca** [comment]
This is ready for review
I've made some small changes to address Rory's concerns.

**Comment by mallenexpensify** [comment]
@johnmlee101 @roryabraham can you review this PR? @kidroca 's been waiting three weeks

**Comment by kidroca** [comment]
Sorry, this accidentally got closed: https://github.com/Expensify/App/pull/4908#issuecomment-1013444052

**Comment by kidroca** [comment]
@johnmlee101 @roryabraham 
Can you post an update
What do I need to do to move this forward?

**Comment by kidroca** [comment]
Oh, I see there are conflicts
I'll be able to merge and test again tomorrow

**Comment by kidroca** [comment]
@johnmlee101 @roryabraham @mallenexpensify 
Before I merge `main`, fix the conflicts and retest everything can someone at least confirm that's the only thing left to do and there are no requested changed

**Comment by johnmlee101** [comment]
I think the changes right now are good, and I'll be ready to approve once the conflicts are resolved.

**Comment by roryabraham** [comment]
Yeah, I looked this over again and I don't think I have any further comments.

**Comment by kidroca** [comment]
Synced with main and retested - everything working as intended

**Review by roryabraham — APPROVED** [review]

File: src/libs/actions/NetworkRequestQueue.js
Comment (sobitneupane) [inline src/libs/actions/NetworkRequestQueue.js:20]: This change has introduced this [bug](https://github.com/Expensify/App/issues/14584). (Title: pinned chat become unpinned when user pinned chat in offline mode)

If a user takes an action that is identical(_.isEqual) to older action then it will be removed.

To fix the issue we have replace above code by:
`
persistedRequests = persistedRequests.concat(requestsToPersist);
`
The PR that fixes the issue: https://github.com/Expensify/App/pull/14608

Comment (kidroca) [reply src/libs/actions/NetworkRequestQueue.js:20]: The idea is to prevent the user from accumulate duplicate requests while offline 
If you perform the same action 10x while offline, we skip posting 10 Network request but just one 
This is an optimization that helps users in low network conditions to not waste bandwidth

From the information shared here it seems that pinning and unpinning the chat uses the same command / request, I would suggest to make them distinct actions so that `_.isEqual` can differentiate them correctly

Comment (iwiznia) [reply src/libs/actions/NetworkRequestQueue.js:20]: We shouldn't have done this. Context https://expensify.slack.com/archives/C01GTK53T8Q/p1685822789307339

**Pull request merged**
